### ISSUE TITLE
no-declaration-in-condition

### ISSUE THREAD

**Issue by johnstonmatt** [eslint/eslint#11342]
**Please describe what the rule should do:**

This rule will prevent accidental true cases when a variable assignment is used in a condition, instead of the conditional check for a truthy condition.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

1)

```js
if(error.code = 'invalid-email') {
  return error;
}

/* will return `error`, when I feel that is unlikely to be the desired behaviour. */

/* the code that exhibits the intended behaviour could be: */

if(error.code === 'invalid-email') {
  return error;
}
```

2)

```js
let indexOfCow = ['dog', 'cat', 'cow'].findIndex((animal) => (animal = 'cow'))
console.log(indexOfCow); // prints 0

let indexOfCow = ['dog', 'cat', 'cow'].findIndex((animal) => (animal === 'cow'))
console.log(indexOfCow); // prints 2

```

**Why should this rule be included in ESLint (instead of a plugin)?**

I think it is a common mistake, a simple solution, and never in my experience the desired behaviour. Its also a very small amount of functionality that would probably be weighed down by extracting it to its own plugin.

**Are you willing to submit a pull request to implement this rule?**
Yes, I have never made a rule, but I would be happy to learn.

**Comment by not-an-aardvark** [comment]
Hi, thanks for the proposal. Have you seen the [`no-cond-assign`](https://eslint.org/docs/rules/no-cond-assign) and [`no-return-assign`](https://eslint.org/docs/rules/no-return-assign) rules? I think those two rules already warn on the types of code you're describing.

**Comment by johnstonmatt** [comment]
I took a look a them just now, I think you are right! I feel silly, eslint didn't save me from a sneaky bug involving this, and my search in the docs didn't come up with anything, so I assumed it didn't exist.
Thanks for your help!
